### ISSUE TITLE
Download all of data.gov 

### ISSUE THREAD

**Issue by flyingzumwalt** [ipfs/archives#113]
Jack Downloads all of data.gov (~350TB) to storage devices on Stanford's network.

**Comment by jonnycrunch** [comment]
Does this really need to be >300TB.  After looking at the data, there is a lot of data redundancy.  Same data is in csv, html and json.   does only one organization have to load the entire 300 TB?    Most of the data can be broken up to 'health', 'environment', "agriculture' and is composed on heterogeneous files ( typically a few hundred MB per file.) The meta data describing the data would be most important ( Publisher, Identifier, modified date, etc).

**Comment by mejackreed** [comment]
We have the ckan metadata already. And yes I agree some of the data is redundant, based on how ArcGIS OpenData allows for different types of exports. A smarter heuristic of this would be nice, but may take some more analysis time.

**Comment by flyingzumwalt** [comment]
@mejackreed do you think you will need help writing the download scripts or running them? We can probably find people to help you.

**Comment by mejackreed** [comment]
Sure thing. Help definitely wanted! I have a naive downloader here: https://github.com/mejackreed/GovScooper/blob/master/README.md#usage already.

**Comment by flyingzumwalt** [comment]
cc @jbenet @gsf @b5

**Comment by b5** [comment]
Happy to help!

I think it makes sense to first decide weather or not to download in passes, using metadata to cut down on data redundancy (as per @jonnycrunch's suggestion), or to just beef the whole thing. I'd personally vote for the "passes" approach, but first checking to ensure that the data is truly redundant.

**Comment by mejackreed** [comment]
Yep i have an idea on how to evaluate whether or not the data is redundant or not. Resources that come from a server that has `/arcgis.com` and have `.geojson` + `.csv` + `.kml` are usually just transformations of the same data. A way to understand these types of datasets / resources and codify the heuristics is needed.

**Comment by jonnycrunch** [comment]
There are 194422 distinct entries in catalog.  Meta data is about 2GB. 

https://catalog.data.gov/api/3/action/package_search?rows=1&start=0

Here is an example of one entry:  
https://catalog.data.gov/api/3/action/package_show?id=1e68f387-5f1c-46c0-a0d1-46044ffef5bf  

each entry has a resource list: 

First pass could be to hit all of the URLs in each resource grab 'Content-Length' headers to calculate the exact amount of space needed simultaneously gathering all of the necessary resource urls.   

There are also some meta schema resources referenced in the 'extras' section that would be important to grab:  https://project-open-data.cio.gov/v1.1/schema/catalog.jsonld

**Comment by jonnycrunch** [comment]
Mmm, at least there were 194422 entries, now there are only 194401.  Now I understand the urgency!

**Comment by b5** [comment]
+1 for hitting all resources for content length. I'd add grabbing filetype while we're at it. quick browsing showed some of the resources listed were .zip archives (ugh)

**Comment by mejackreed** [comment]
So in my initial tests of downloading these resources, many of them do not return Content-Length header unfortunately. Hoping to kick off some larger runs this afternoon to get more details.

**Comment by mejackreed** [comment]
@jonnycrunch 194014 entries here: https://github.com/OpenGeoMetadata/gov.data

Best to grab the archive.zip and easy to parse the layers.json file
